diff --git a/lib/resty/websocket/client.lua b/lib/resty/websocket/client.lua
index e24cc05..dc4e1f2 100644
--- a/lib/resty/websocket/client.lua
+++ b/lib/resty/websocket/client.lua
@@ -32,7 +32,8 @@ local ssl_support = true
 
 if not ngx.config
     or not ngx.config.ngx_lua_version
-    or ngx.config.ngx_lua_version < 9011
+    or (ngx.config.subsystem ~= "stream"
+        and ngx.config.ngx_lua_version < 9011)
 then
     ssl_support = false
 end
