diff --git a/config b/config
index 56c7b55..d541c25 100644
--- a/config
+++ b/config
@@ -282,6 +282,7 @@ STREAM_LUA_SRCS=" \
                 $ngx_addon_dir/src/ngx_stream_lua_proxy_ssl_verifyby.c \
                 $ngx_addon_dir/src/ngx_stream_lua_log_ringbuf.c \
                 $ngx_addon_dir/src/ngx_stream_lua_input_filters.c \
+                $ngx_addon_dir/src/ngx_stream_lua_accessby.c \
                 "
 
 STREAM_LUA_DEPS=" \
@@ -328,6 +329,7 @@ STREAM_LUA_DEPS=" \
                 $ngx_addon_dir/src/ngx_stream_lua_proxy_ssl_verifyby.h \
                 $ngx_addon_dir/src/ngx_stream_lua_log_ringbuf.h \
                 $ngx_addon_dir/src/ngx_stream_lua_input_filters.h \
+                $ngx_addon_dir/src/ngx_stream_lua_accessby.h \
                 "
 
 # ----------------------------------------
diff --git a/src/ngx_stream_lua_accessby.c b/src/ngx_stream_lua_accessby.c
new file mode 100644
index 0000000..e910f31
--- /dev/null
+++ b/src/ngx_stream_lua_accessby.c
@@ -0,0 +1,326 @@
+
+/*
+ * Copyright (C) Hanada <im@hanada.info>
+ * Copyright (C) Arcadiy Ivanov (arcivanov)
+ * Copyright (C) OpenResty Inc.
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include <nginx.h>
+#include "ngx_stream_lua_accessby.h"
+#include "ngx_stream_lua_util.h"
+#include "ngx_stream_lua_exception.h"
+#include "ngx_stream_lua_cache.h"
+
+
+static ngx_int_t ngx_stream_lua_access_by_chunk(lua_State *L,
+    ngx_stream_lua_request_t *r);
+
+
+ngx_int_t
+ngx_stream_lua_access_handler(ngx_stream_session_t *s)
+{
+    ngx_int_t                     rc;
+    ngx_stream_lua_ctx_t         *ctx;
+    ngx_stream_lua_srv_conf_t    *lscf;
+    ngx_stream_lua_main_conf_t   *lmcf;
+    ngx_stream_lua_request_t     *r;
+
+    ngx_log_debug0(NGX_LOG_DEBUG_STREAM, s->connection->log, 0,
+                   "lua access handler");
+
+    lmcf = ngx_stream_get_module_main_conf(s, ngx_stream_lua_module);
+
+    if (!lmcf->postponed_to_access_phase_end) {
+        ngx_stream_core_main_conf_t       *cmcf;
+        ngx_stream_phase_handler_t         tmp;
+        ngx_stream_phase_handler_t        *ph;
+        ngx_stream_phase_handler_t        *cur_ph;
+        ngx_stream_phase_handler_t        *last_ph;
+
+        lmcf->postponed_to_access_phase_end = 1;
+
+        cmcf = ngx_stream_get_module_main_conf(s, ngx_stream_core_module);
+
+        ph = cmcf->phase_engine.handlers;
+        cur_ph = &ph[s->phase_handler];
+        last_ph = &ph[cur_ph->next - 1];
+
+        if (cur_ph < last_ph) {
+            tmp      = *cur_ph;
+
+            ngx_memmove(cur_ph, cur_ph + 1, (last_ph - cur_ph)
+                        * sizeof (ngx_stream_phase_handler_t));
+
+            *last_ph = tmp;
+
+            s->phase_handler--; /* redo the current ph */
+
+            return NGX_DECLINED;
+        }
+    }
+
+    lscf = ngx_stream_get_module_srv_conf(s, ngx_stream_lua_module);
+
+    if (lscf->access_handler == NULL) {
+        ngx_log_debug0(NGX_LOG_DEBUG_STREAM, s->connection->log, 0,
+                       "no acces handler found");
+        return NGX_DECLINED;
+    }
+
+    ctx = ngx_stream_get_module_ctx(s, ngx_stream_lua_module);
+
+    dd("ctx = %p", ctx);
+
+    if (ctx == NULL) {
+        ctx = ngx_stream_lua_create_ctx(s);
+        if (ctx == NULL) {
+            return NGX_STREAM_INTERNAL_SERVER_ERROR;
+        }
+    }
+
+    r = ctx->request;
+
+    dd("entered? %d", (int) ctx->entered_access_phase);
+
+    if (ctx->entered_access_phase) {
+        dd("calling wev handler");
+        rc = ctx->resume_handler(r);
+        dd("wev handler returns %d", (int) rc);
+
+        if (rc == NGX_ERROR || rc > NGX_OK) {
+            ngx_stream_lua_finalize_request(ctx->request, rc);
+            return NGX_DONE;
+        }
+
+        if (rc == NGX_OK || rc == NGX_DONE) {
+            return rc;
+        }
+
+        return NGX_DECLINED;
+    }
+
+    r->connection->read->handler = ngx_stream_lua_request_handler;
+    r->connection->write->handler = ngx_stream_lua_request_handler;
+
+    dd("calling access handler");
+    rc = lscf->access_handler(r);
+
+    if (rc == NGX_ERROR || rc > NGX_OK) {
+        ngx_stream_lua_finalize_request(ctx->request, rc);
+        return NGX_DONE;
+    }
+
+    return rc;
+}
+
+
+ngx_int_t
+ngx_stream_lua_access_handler_inline(ngx_stream_lua_request_t *r)
+{
+    ngx_int_t                    rc;
+    lua_State                   *L;
+    ngx_stream_lua_srv_conf_t   *lscf;
+
+    lscf = ngx_stream_lua_get_module_srv_conf(r, ngx_stream_lua_module);
+
+    L = ngx_stream_lua_get_lua_vm(r, NULL);
+
+    /*  load Lua inline script (w/ cache) sp = 1 */
+    rc = ngx_stream_lua_cache_loadbuffer(r->connection->log, L,
+                                         lscf->access_src.value.data,
+                                         lscf->access_src.value.len,
+                                         lscf->access_src_key,
+                                         (const char *) lscf->access_chunkname);
+
+    if (rc != NGX_OK) {
+        return NGX_STREAM_INTERNAL_SERVER_ERROR;
+    }
+
+    return ngx_stream_lua_access_by_chunk(L, r);
+}
+
+
+ngx_int_t
+ngx_stream_lua_access_handler_file(ngx_stream_lua_request_t *r)
+{
+    u_char                      *script_path;
+    ngx_int_t                    rc;
+    ngx_str_t                    eval_src;
+    lua_State                   *L;
+    ngx_stream_lua_srv_conf_t   *lscf;
+
+    lscf = ngx_stream_lua_get_module_srv_conf(r, ngx_stream_lua_module);
+
+    /* Eval nginx variables in code path string first */
+    if (ngx_stream_complex_value(r->session, &lscf->access_src, &eval_src)
+        != NGX_OK)
+    {
+        return NGX_ERROR;
+    }
+
+    script_path = ngx_stream_lua_rebase_path(r->pool, eval_src.data,
+                                             eval_src.len);
+
+    if (script_path == NULL) {
+        return NGX_ERROR;
+    }
+
+    L = ngx_stream_lua_get_lua_vm(r, NULL);
+
+    /*  load Lua script file (w/ cache)        sp = 1 */
+    rc = ngx_stream_lua_cache_loadfile(r->connection->log, L, script_path,
+                                       lscf->access_src_key);
+    if (rc != NGX_OK) {
+        return rc;
+    }
+
+    /*  make sure we have a valid code chunk */
+    ngx_stream_lua_assert(lua_isfunction(L, -1));
+
+    return ngx_stream_lua_access_by_chunk(L, r);
+}
+
+
+static ngx_int_t
+ngx_stream_lua_access_by_chunk(lua_State *L, ngx_stream_lua_request_t *r)
+{
+    int                          co_ref;
+    ngx_int_t                    rc;
+    lua_State                   *co;
+    ngx_event_t                 *rev;
+    ngx_connection_t            *c;
+    ngx_stream_lua_ctx_t        *ctx;
+    ngx_stream_lua_cleanup_t    *cln;
+
+    ngx_stream_lua_srv_conf_t     *lscf;
+
+    /*  {{{ new coroutine to handle request */
+    co = ngx_stream_lua_new_thread(r, L, &co_ref);
+
+    if (co == NULL) {
+        ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                      "lua: failed to create new coroutine "
+                      "to handle request");
+
+        return NGX_STREAM_INTERNAL_SERVER_ERROR;
+    }
+
+    /*  move code closure to new coroutine */
+    lua_xmove(L, co, 1);
+
+#ifndef OPENRESTY_LUAJIT
+    /*  set closure's env table to new coroutine's globals table */
+    ngx_stream_lua_get_globals_table(co);
+    lua_setfenv(co, -2);
+#endif
+
+    /*  save nginx request in coroutine globals table */
+    ngx_stream_lua_set_req(co, r);
+
+    /*  {{{ initialize request context */
+    ctx = ngx_stream_lua_get_module_ctx(r, ngx_stream_lua_module);
+
+    dd("ctx = %p", ctx);
+
+    if (ctx == NULL) {
+        return NGX_ERROR;
+    }
+
+    ngx_stream_lua_reset_ctx(r, L, ctx);
+
+    ctx->entered_access_phase = 1;
+
+    ctx->cur_co_ctx = &ctx->entry_co_ctx;
+    ctx->cur_co_ctx->co = co;
+    ctx->cur_co_ctx->co_ref = co_ref;
+#ifdef NGX_LUA_USE_ASSERT
+    ctx->cur_co_ctx->co_top = 1;
+#endif
+
+    ngx_stream_lua_attach_co_ctx_to_L(co, ctx->cur_co_ctx);
+
+    /*  }}} */
+
+    /*  {{{ register request cleanup hooks */
+    if (ctx->cleanup == NULL) {
+        cln = ngx_stream_lua_cleanup_add(r, 0);
+        if (cln == NULL) {
+            return NGX_STREAM_INTERNAL_SERVER_ERROR;
+        }
+
+        cln->handler = ngx_stream_lua_request_cleanup_handler;
+        cln->data = ctx;
+        ctx->cleanup = &cln->handler;
+    }
+    /*  }}} */
+
+    ctx->context = NGX_STREAM_LUA_CONTEXT_ACCESS;
+
+    lscf = ngx_stream_lua_get_module_srv_conf(r, ngx_stream_lua_module);
+
+    if (lscf->check_client_abort) {
+        r->read_event_handler = ngx_stream_lua_rd_check_broken_connection;
+
+        rev = r->connection->read;
+
+        if (!rev->active) {
+            if (ngx_add_event(rev, NGX_READ_EVENT, 0) != NGX_OK) {
+                return NGX_ERROR;
+            }
+        }
+
+    } else {
+        r->read_event_handler = ngx_stream_lua_block_reading;
+    }
+
+    rc = ngx_stream_lua_run_thread(L, r, ctx, 0);
+
+    ngx_log_debug1(NGX_LOG_DEBUG_STREAM, r->connection->log, 0,
+                   "access run thread returned %d", (int) rc);
+
+    if (rc == NGX_ERROR || rc > NGX_OK) {
+        return rc;
+    }
+
+    c = r->connection;
+
+    if (rc == NGX_AGAIN) {
+        rc = ngx_stream_lua_run_posted_threads(c, L, r, ctx, 0);
+
+        if (rc == NGX_ERROR || rc == NGX_DONE || rc > NGX_OK) {
+            return rc;
+        }
+
+        if (rc != NGX_OK) {
+            return NGX_DECLINED;
+        }
+
+    } else if (rc == NGX_DONE) {
+        ngx_stream_lua_finalize_request(r, NGX_DONE);
+
+        rc = ngx_stream_lua_run_posted_threads(c, L, r, ctx, 0);
+
+        if (rc == NGX_ERROR || rc == NGX_DONE || rc > NGX_OK) {
+            return rc;
+        }
+
+        if (rc != NGX_OK) {
+            return NGX_DECLINED;
+        }
+    }
+
+    if (rc == NGX_OK) {
+        return NGX_OK;
+    }
+
+    return NGX_DECLINED;
+}
+
+/* vi:set ft=c ts=4 sw=4 et fdm=marker: */
diff --git a/src/ngx_stream_lua_accessby.h b/src/ngx_stream_lua_accessby.h
new file mode 100644
index 0000000..dbd4f13
--- /dev/null
+++ b/src/ngx_stream_lua_accessby.h
@@ -0,0 +1,23 @@
+
+/*
+ * Copyright (C) Hanada <im@hanada.info>
+ * Copyright (C) Arcadiy Ivanov (arcivanov)
+ * Copyright (C) OpenResty Inc.
+ */
+
+
+#ifndef _NGX_STREAM_LUA_ACCESSBY_H_INCLUDED_
+#define _NGX_STREAM_LUA_ACCESSBY_H_INCLUDED_
+
+
+#include "ngx_stream_lua_common.h"
+
+
+ngx_int_t ngx_stream_lua_access_handler(ngx_stream_session_t *s);
+ngx_int_t ngx_stream_lua_access_handler_inline(ngx_stream_lua_request_t *r);
+ngx_int_t ngx_stream_lua_access_handler_file(ngx_stream_lua_request_t *r);
+
+
+#endif /* _NGX_STREAM_LUA_ACCESSBY_H_INCLUDED_ */
+
+/* vi:set ft=c ts=4 sw=4 et fdm=marker: */
diff --git a/src/ngx_stream_lua_common.h b/src/ngx_stream_lua_common.h
index 1126f7d..5d84595 100644
--- a/src/ngx_stream_lua_common.h
+++ b/src/ngx_stream_lua_common.h
@@ -154,6 +154,7 @@
 #define NGX_STREAM_LUA_CONTEXT_SSL_CLIENT_HELLO                     0x0080
 #define NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY                     0x0100
 #define NGX_STREAM_LUA_CONTEXT_PROXY_SSL_CERT                       0x0200
+#define NGX_STREAM_LUA_CONTEXT_ACCESS                               0x0400
 
 
 #define NGX_STREAM_LUA_FFI_NO_REQ_CTX         -100
@@ -232,6 +233,7 @@ struct ngx_stream_lua_main_conf_s {
     ngx_array_t         *preload_hooks; /* of ngx_stream_lua_preload_hook_t */
 
     ngx_flag_t           postponed_to_preread_phase_end;
+    ngx_flag_t           postponed_to_access_phase_end;
 
     ngx_stream_lua_main_conf_handler_pt          init_handler;
     ngx_str_t                                    init_src;
@@ -258,7 +260,7 @@ struct ngx_stream_lua_main_conf_s {
     ngx_flag_t           set_sa_restart;
 
     unsigned             requires_preread:1;
-
+    unsigned             requires_access:1;
     unsigned             requires_log:1;
     unsigned             requires_shm:1;
     unsigned             requires_capture_log:1;
@@ -314,6 +316,11 @@ struct ngx_stream_lua_srv_conf_s {
     ngx_stream_complex_value_t          preread_src;
     u_char                             *preread_src_key;
 
+    ngx_stream_lua_handler_pt           access_handler;
+    u_char                             *access_chunkname;
+    ngx_stream_complex_value_t          access_src;
+    u_char                             *access_src_key;
+
     ngx_stream_lua_handler_pt           content_handler;
     u_char                             *content_chunkname;
     ngx_stream_complex_value_t          content_src;
@@ -518,7 +525,7 @@ typedef struct ngx_stream_lua_ctx_s {
                                        response headers */
 
     unsigned         entered_preread_phase:1;
-
+    unsigned         entered_access_phase:1;
     unsigned         entered_content_phase:1;
 
     unsigned         buffering:1; /* HTTP 1.0 response body buffering flag */
diff --git a/src/ngx_stream_lua_control.c b/src/ngx_stream_lua_control.c
index 6d9089c..65dbf49 100644
--- a/src/ngx_stream_lua_control.c
+++ b/src/ngx_stream_lua_control.c
@@ -120,7 +120,8 @@ ngx_stream_lua_ffi_exit(ngx_stream_lua_request_t *r, int status, u_char *err,
         | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_CERT
         | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY
 #endif
-        | NGX_STREAM_LUA_CONTEXT_PREREAD,
+        | NGX_STREAM_LUA_CONTEXT_PREREAD
+        | NGX_STREAM_LUA_CONTEXT_ACCESS,
         err, errlen) != NGX_OK)
     {
         return NGX_ERROR;
diff --git a/src/ngx_stream_lua_coroutine.c b/src/ngx_stream_lua_coroutine.c
index a8533ca..d97df3e 100644
--- a/src/ngx_stream_lua_coroutine.c
+++ b/src/ngx_stream_lua_coroutine.c
@@ -210,6 +210,7 @@ ngx_stream_lua_coroutine_resume(lua_State *L)
                                  | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY
 #endif
                                  | NGX_STREAM_LUA_CONTEXT_PREREAD
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS
                                  );
 
     p_coctx = ctx->cur_co_ctx;
@@ -275,6 +276,7 @@ ngx_stream_lua_coroutine_yield(lua_State *L)
                                  | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY
 #endif
                                  | NGX_STREAM_LUA_CONTEXT_PREREAD
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS
                                  );
 
     coctx = ctx->cur_co_ctx;
@@ -439,6 +441,7 @@ ngx_stream_lua_coroutine_status(lua_State *L)
                                  | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY
 #endif
                                  | NGX_STREAM_LUA_CONTEXT_PREREAD
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS
                                  );
 
     coctx = ngx_stream_lua_get_co_ctx(co, ctx);
diff --git a/src/ngx_stream_lua_directive.c b/src/ngx_stream_lua_directive.c
index 61519e9..fbffa5e 100644
--- a/src/ngx_stream_lua_directive.c
+++ b/src/ngx_stream_lua_directive.c
@@ -34,6 +34,7 @@
 #include "api/ngx_stream_lua_api.h"
 
 #include "ngx_stream_lua_prereadby.h"
+#include "ngx_stream_lua_accessby.h"
 
 
 typedef struct ngx_stream_lua_block_parser_ctx_s
@@ -234,6 +235,121 @@ ngx_stream_lua_package_path(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 }
 
 
+char *
+ngx_stream_lua_access_by_lua_block(ngx_conf_t *cf, ngx_command_t *cmd,
+    void *conf)
+{
+    char        *rv;
+    ngx_conf_t   save;
+
+    save = *cf;
+    cf->handler = ngx_stream_lua_access_by_lua;
+    cf->handler_conf = conf;
+
+    rv = ngx_stream_lua_conf_lua_block_parse(cf, cmd);
+
+    *cf = save;
+
+    return rv;
+}
+
+
+char *
+ngx_stream_lua_access_by_lua(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+{
+    size_t                                 chunkname_len;
+    u_char                                *p, *chunkname;
+    ngx_str_t                             *value;
+    ngx_stream_lua_main_conf_t            *lmcf;
+    ngx_stream_lua_srv_conf_t             *lscf = conf;
+
+    ngx_stream_compile_complex_value_t     ccv;
+
+    dd("enter");
+
+    /*  must specify an access handler */
+    if (cmd->post == NULL) {
+        return NGX_CONF_ERROR;
+    }
+
+    if (lscf->access_handler) {
+        return "is duplicate";
+    }
+
+    value = cf->args->elts;
+
+    if (value[1].len == 0) {
+        /*  Oops...Invalid server conf */
+        ngx_conf_log_error(NGX_LOG_ERR, cf, 0,
+                           "invalid server config: no runnable Lua code");
+
+        return NGX_CONF_ERROR;
+    }
+
+    if (cmd->post == ngx_stream_lua_access_handler_inline) {
+        chunkname = ngx_stream_lua_gen_chunk_name(cf, "access_by_lua",
+                                                  sizeof("access_by_lua") - 1,
+                                                  &chunkname_len);
+        if (chunkname == NULL) {
+            return NGX_CONF_ERROR;
+        }
+
+        lscf->access_chunkname = chunkname;
+
+        /* Don't eval nginx variables for inline lua code */
+
+        lscf->access_src.value = value[1];
+
+        p = ngx_palloc(cf->pool,
+                       chunkname_len + NGX_STREAM_LUA_INLINE_KEY_LEN + 1);
+        if (p == NULL) {
+            return NGX_CONF_ERROR;
+        }
+
+        lscf->access_src_key = p;
+
+        p = ngx_copy(p, chunkname, chunkname_len);
+        p = ngx_copy(p, NGX_STREAM_LUA_INLINE_TAG,
+                     NGX_STREAM_LUA_INLINE_TAG_LEN);
+        p = ngx_stream_lua_digest_hex(p, value[1].data, value[1].len);
+        *p = '\0';
+
+    } else {
+        ngx_memzero(&ccv, sizeof(ngx_stream_compile_complex_value_t));
+        ccv.cf = cf;
+        ccv.value = &value[1];
+        ccv.complex_value = &lscf->access_src;
+
+        if (ngx_stream_compile_complex_value(&ccv) != NGX_OK) {
+            return NGX_CONF_ERROR;
+        }
+
+        if (lscf->access_src.lengths == NULL) {
+            /* no variable found */
+            p = ngx_palloc(cf->pool, NGX_STREAM_LUA_FILE_KEY_LEN + 1);
+            if (p == NULL) {
+                return NGX_CONF_ERROR;
+            }
+
+            lscf->access_src_key = p;
+
+            p = ngx_copy(p, NGX_STREAM_LUA_FILE_TAG,
+                         NGX_STREAM_LUA_FILE_TAG_LEN);
+            p = ngx_stream_lua_digest_hex(p, value[1].data, value[1].len);
+            *p = '\0';
+        }
+    }
+
+    lscf->access_handler = (ngx_stream_lua_handler_pt) cmd->post;
+
+    lmcf = ngx_stream_conf_get_module_main_conf(cf, ngx_stream_lua_module);
+
+    lmcf->requires_access = 1;
+
+    return NGX_CONF_OK;
+}
+
+
 char *
 ngx_stream_lua_preread_by_lua_block(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf)
diff --git a/src/ngx_stream_lua_directive.h b/src/ngx_stream_lua_directive.h
index 6b36bb1..3fce692 100644
--- a/src/ngx_stream_lua_directive.h
+++ b/src/ngx_stream_lua_directive.h
@@ -54,6 +54,10 @@ ngx_stream_lua_preread_by_lua_block(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
 char *
 ngx_stream_lua_preread_by_lua(ngx_conf_t *cf, ngx_command_t *cmd, void *conf);
+char *ngx_stream_lua_access_by_lua_block(ngx_conf_t *cf, ngx_command_t *cmd,
+    void *conf);
+char *ngx_stream_lua_access_by_lua(ngx_conf_t *cf, ngx_command_t *cmd,
+    void *conf);
 char *
 ngx_stream_lua_add_variable(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
diff --git a/src/ngx_stream_lua_module.c b/src/ngx_stream_lua_module.c
index 772e8c5..1cbb3a6 100644
--- a/src/ngx_stream_lua_module.c
+++ b/src/ngx_stream_lua_module.c
@@ -37,6 +37,7 @@
 #endif
 
 #include "ngx_stream_lua_prereadby.h"
+#include "ngx_stream_lua_accessby.h"
 
 
 #if !defined(OPENSSL_VERSION_NUMBER) || OPENSSL_VERSION_NUMBER > 0x101010afL
@@ -256,6 +257,28 @@ static ngx_command_t ngx_stream_lua_cmds[] = {
       0,
       (void *) ngx_stream_lua_preread_handler_inline },
 
+    /* access_by_lua_block { <inline script> } */
+    { ngx_string("access_by_lua_block"),
+      NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_BLOCK|NGX_CONF_NOARGS,
+      ngx_stream_lua_access_by_lua_block,
+      NGX_STREAM_SRV_CONF_OFFSET,
+      0,
+      (void *) ngx_stream_lua_access_handler_inline },
+
+    /* access_by_lua_file rel/or/abs/path/to/script */
+    { ngx_string("access_by_lua_file"),
+      NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_stream_lua_access_by_lua,
+      NGX_STREAM_SRV_CONF_OFFSET,
+      0,
+      (void *) ngx_stream_lua_access_handler_file },
+
+    { ngx_string("access_by_lua_no_postpone"),
+      NGX_STREAM_MAIN_CONF|NGX_CONF_FLAG,
+      ngx_conf_set_flag_slot,
+      NGX_STREAM_MAIN_CONF_OFFSET,
+      offsetof(ngx_stream_lua_main_conf_t, postponed_to_access_phase_end),
+      NULL },
 
     /* content_by_lua "<inline script>" */
     { ngx_string("content_by_lua"),
@@ -607,6 +630,21 @@ ngx_stream_lua_init(ngx_conf_t *cf)
         lmcf->postponed_to_preread_phase_end = 0;
     }
 
+    dd("requires access: %d", (int) lmcf->requires_access);
+
+    if (lmcf->requires_access) {
+        h = ngx_array_push(&cmcf->phases[NGX_STREAM_ACCESS_PHASE].handlers);
+        if (h == NULL) {
+            return NGX_ERROR;
+        }
+
+        *h = ngx_stream_lua_access_handler;
+    }
+
+    if (lmcf->postponed_to_access_phase_end == NGX_CONF_UNSET) {
+        lmcf->postponed_to_access_phase_end = 0;
+    }
+
     dd("requires log: %d", (int) lmcf->requires_log);
 
     if (lmcf->requires_log) {
@@ -802,6 +840,7 @@ ngx_stream_lua_create_main_conf(ngx_conf_t *cf)
 #endif
 
     lmcf->postponed_to_preread_phase_end = NGX_CONF_UNSET;
+    lmcf->postponed_to_access_phase_end = NGX_CONF_UNSET;
 
     lmcf->set_sa_restart = NGX_CONF_UNSET;
 
@@ -1127,6 +1166,13 @@ ngx_stream_lua_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
         conf->preread_chunkname = prev->preread_chunkname;
     }
 
+    if (conf->access_src.value.len == 0) {
+        conf->access_src = prev->access_src;
+        conf->access_handler = prev->access_handler;
+        conf->access_src_key = prev->access_src_key;
+        conf->access_chunkname = prev->access_chunkname;
+    }
+
     if (conf->log_src.value.len == 0) {
         conf->log_src = prev->log_src;
         conf->log_handler = prev->log_handler;
diff --git a/src/ngx_stream_lua_output.c b/src/ngx_stream_lua_output.c
index 2e5317e..a059148 100644
--- a/src/ngx_stream_lua_output.c
+++ b/src/ngx_stream_lua_output.c
@@ -76,7 +76,8 @@ ngx_stream_lua_ngx_echo(lua_State *L, unsigned newline)
     }
 
     ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT
-                                 | NGX_STREAM_LUA_CONTEXT_PREREAD);
+                                 | NGX_STREAM_LUA_CONTEXT_PREREAD
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS);
 
 
     if (ctx->eof) {
@@ -496,7 +497,8 @@ ngx_stream_lua_ngx_flush(lua_State *L)
     }
 
     ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT
-                                 | NGX_STREAM_LUA_CONTEXT_PREREAD);
+                                 | NGX_STREAM_LUA_CONTEXT_PREREAD
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS);
 
 
     coctx = ctx->cur_co_ctx;
@@ -619,7 +621,8 @@ ngx_stream_lua_ngx_eof(lua_State *L)
     }
 
     ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT
-                                 | NGX_STREAM_LUA_CONTEXT_PREREAD);
+                                 | NGX_STREAM_LUA_CONTEXT_PREREAD
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS);
 
     ngx_log_debug0(NGX_LOG_DEBUG_STREAM, r->connection->log, 0,
                    "lua send eof");
diff --git a/src/ngx_stream_lua_phase.c b/src/ngx_stream_lua_phase.c
index 35bbfe5..42e1288 100644
--- a/src/ngx_stream_lua_phase.c
+++ b/src/ngx_stream_lua_phase.c
@@ -62,6 +62,10 @@ ngx_stream_lua_ngx_get_phase(lua_State *L)
         lua_pushliteral(L, "preread");
         break;
 
+    case NGX_STREAM_LUA_CONTEXT_ACCESS:
+        lua_pushliteral(L, "access");
+        break;
+
     case NGX_STREAM_LUA_CONTEXT_CONTENT:
         lua_pushliteral(L, "content");
         break;
diff --git a/src/ngx_stream_lua_socket_tcp.c b/src/ngx_stream_lua_socket_tcp.c
index e579d59..e5b3e5e 100644
--- a/src/ngx_stream_lua_socket_tcp.c
+++ b/src/ngx_stream_lua_socket_tcp.c
@@ -5865,7 +5865,8 @@ ngx_stream_lua_req_socket_tcp(lua_State *L)
     }
 
     ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT
-                                 |NGX_STREAM_LUA_CONTEXT_PREREAD);
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS
+                                 | NGX_STREAM_LUA_CONTEXT_PREREAD);
 
     c = r->connection;
 
diff --git a/src/ngx_stream_lua_socket_udp.c b/src/ngx_stream_lua_socket_udp.c
index b0ed4d1..57ebea4 100644
--- a/src/ngx_stream_lua_socket_udp.c
+++ b/src/ngx_stream_lua_socket_udp.c
@@ -2093,6 +2093,7 @@ ngx_stream_lua_req_socket_udp(lua_State *L)
     }
 
     ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT
+                                 | NGX_STREAM_LUA_CONTEXT_ACCESS
                                  |NGX_STREAM_LUA_CONTEXT_PREREAD);
 
     c = r->connection;
diff --git a/src/ngx_stream_lua_uthread.c b/src/ngx_stream_lua_uthread.c
index e91baca..21d18fc 100644
--- a/src/ngx_stream_lua_uthread.c
+++ b/src/ngx_stream_lua_uthread.c
@@ -232,7 +232,7 @@ ngx_stream_lua_uthread_kill(lua_State *L)
     }
 
     ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT
-
+                               | NGX_STREAM_LUA_CONTEXT_ACCESS
                                | NGX_STREAM_LUA_CONTEXT_PREREAD
                                | NGX_STREAM_LUA_CONTEXT_SSL_CLIENT_HELLO
                                | NGX_STREAM_LUA_CONTEXT_SSL_CERT
diff --git a/src/ngx_stream_lua_util.h b/src/ngx_stream_lua_util.h
index e395ecc..6cd1929 100644
--- a/src/ngx_stream_lua_util.h
+++ b/src/ngx_stream_lua_util.h
@@ -78,7 +78,8 @@ extern char ngx_stream_lua_headers_metatable_key;
                                 | NGX_STREAM_LUA_CONTEXT_SSL_CLIENT_HELLO    \
                                 | NGX_STREAM_LUA_CONTEXT_SSL_CERT            \
                                 | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_CERT      \
-                                | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY)
+                                | NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY    \
+                                | NGX_STREAM_LUA_CONTEXT_ACCESS)
 
 
 #define ngx_stream_lua_context_name(c)                                       \
@@ -95,4 +96,5 @@ extern char ngx_stream_lua_headers_metatable_key;
                                             "proxy_ssl_certificate_by_lua*"  \
      : (c) == NGX_STREAM_LUA_CONTEXT_PROXY_SSL_VERIFY ?                      \
                                                  "proxy_ssl_verify_by_lua*"  \
+     : (c) == NGX_STREAM_LUA_CONTEXT_ACCESS ? "access_by_lua*"               \
      : "(unknown)")
