diff --git a/src/nginx/error.cpp b/src/nginx/error.cpp
index 0193a41..7779f00 100644
--- a/src/nginx/error.cpp
+++ b/src/nginx/error.cpp
@@ -49,7 +49,7 @@ ngx_int_t ngx_weserv_return_error(ngx_http_request_t *r,
     buf->last_in_chain = 1;
     buf->last = ngx_cpymem(buf->last, error.c_str(), content_length);
 
-    r->headers_out.status = http_status;
+    r->headers_out.status = NGX_HTTP_INTERNAL_SERVER_ERROR;
     r->headers_out.content_type_len = sizeof("application/json") - 1;
     ngx_str_set(&r->headers_out.content_type, "application/json");
     r->headers_out.content_type_lowcase = nullptr;
