diff --git a/ngx_rtmp_codec_module.c b/ngx_rtmp_codec_module.c
index 20e8d40..731a0de 100644
--- a/ngx_rtmp_codec_module.c
+++ b/ngx_rtmp_codec_module.c
@@ -625,8 +625,8 @@ ngx_rtmp_codec_reconstruct_meta(ngx_rtmp_session_t *s)
     static ngx_rtmp_amf_elt_t       out_inf[] = {
 
         { NGX_RTMP_AMF_STRING,
-          ngx_string("Server"),
-          "NGINX HTTP-FLV (https://github.com/winshining/nginx-http-flv-module)", 0 },
+          ngx_string("server"),
+          "openresty", 0 },
 
         { NGX_RTMP_AMF_NUMBER,
           ngx_string("width"),
